ssl/quic/quic_channel.c

```diff
@@ -0,0 +1,1627 @@
+{
+    QUIC_CHANNEL *ch = arg;
+
+    return uni ? ch->max_local_streams_uni : ch->max_local_streams_bidi;
+}
+
+/*
+ * Called by QRX to determine if a packet is potentially invalid before trying
+ * to decrypt it.
+ */
+static int rx_early_validate(QUIC_PN pn, int pn_space, void *arg)
+{
+    QUIC_CHANNEL *ch = arg;
+
+    /* Potential duplicates should not be processed. */
+    if (!ossl_ackm_is_rx_pn_processable(ch->ackm, pn, pn_space))
+        return 0;
+
+    return 1;
+}
+
+/*
+ * QUIC Channel: Handshake Layer Event Handling
+ * ============================================
+ */
+static int ch_on_crypto_send(const unsigned char *buf, size_t buf_len,
+                             size_t *consumed, void *arg)
+{
+    int ret;
+    QUIC_CHANNEL *ch = arg;
+    uint32_t enc_level = ch->tx_enc_level;
+    uint32_t pn_space = ossl_quic_enc_level_to_pn_space(enc_level);
+    QUIC_SSTREAM *sstream = ch->crypto_send[pn_space];
+
+    if (!ossl_assert(sstream != NULL))
+        return 0;
+
+    ret = ossl_quic_sstream_append(sstream, buf, buf_len, consumed);
+    return ret;
+}
+
+static int crypto_ensure_empty(QUIC_RSTREAM *rstream)
+{
+    size_t avail = 0;
+    int is_fin = 0;
+
+    if (rstream == NULL)
+        return 1;
+
+    if (!ossl_quic_rstream_available(rstream, &avail, &is_fin))
+        return 0;
+
+    return avail == 0;
+}
+
+static int ch_on_crypto_recv(unsigned char *buf, size_t buf_len,
+                             size_t *bytes_read, void *arg)
+{
+    QUIC_CHANNEL *ch = arg;
+    QUIC_RSTREAM *rstream;
+    int is_fin = 0; /* crypto stream is never finished, so we don't use this */
+    uint32_t i;
+
+    /*
+     * After we move to a later EL we must not allow our peer to send any new
+     * bytes in the crypto stream on a previous EL. Retransmissions of old bytes
+     * are allowed.
+     *
+     * In practice we will only move to a new EL when we have consumed all bytes
+     * which should be sent on the crypto stream at a previous EL. For example,
+     * the Handshake EL should not be provisioned until we have completely
+     * consumed a TLS 1.3 ServerHello. Thus when we provision an EL the output
+     * of ossl_quic_rstream_available() should be 0 for all lower ELs. Thus if a
+     * given EL is available we simply ensure we have not received any further
+     * bytes at a lower EL.
+     */
+    for (i = QUIC_ENC_LEVEL_INITIAL; i < ch->tx_enc_level; ++i)
+        if (i != QUIC_ENC_LEVEL_0RTT &&
+            !crypto_ensure_empty(ch->crypto_recv[ossl_quic_enc_level_to_pn_space(i)])) {
+            /* Protocol violation (RFC 9001 s. 4.1.3) */
+            ossl_quic_channel_raise_protocol_error(ch, QUIC_ERR_PROTOCOL_VIOLATION,
+                                                   OSSL_QUIC_FRAME_TYPE_CRYPTO,
+                                                   "crypto stream data in wrong EL");
+            return 0;
+        }
+
+    rstream = ch->crypto_recv[ossl_quic_enc_level_to_pn_space(ch->tx_enc_level)];
```

Comment:
We seem to be deciding which rstream to *receive* on based on the *tx* enc level. This looks wrong. As a client we update the rx secrets when we receive the ServerHello. We don't update the tx secrets until just before we send the Client Certificate/CertificateVerify or Finished. So to the tx level is behind the rx level. This means we don't find the data we are expecting and we stall.

Reply:
Wait, what? Is there some reason we aren't provisioning TX and RX secrets at the same time? Once the client has a SH we shouldn't have any reason not to provision both.

Reply:
I am fairly are some corner cases related to early_data that mean we only want to change the tx secret at the last possible moment - but we need to change the rx secret much earlier so that we can process the incoming CRYPTO data. I am very hesitant to change this in the TLS stack. It seems that just having an equivalent rx_enc_level would be sufficient to solve this problem.